csaf_2.1/prose/edit/src/schema-elements-02-props-03-vulnerabilities.md

```diff
@@ -15,7 +15,7 @@ properties represents a list of all relevant vulnerability information items.
 The Vulnerability item of value type `object` with 1 or more properties is a container for the aggregation of all fields that are related to
 a single vulnerability in the document.
 Any vulnerability MAY provide the optional properties Acknowledgments (`acknowledgments`), Common Vulnerabilities and Exposures (CVE) (`cve`),
-Common Weakness Enumeration (CWE) (`cwe`), Discovery Date (`discovery_date`), Flags (`flags`), IDs (`ids`), Involvements (`involvements`),
+Common Weakness Enumerations (CWEs) (`cwes`), Discovery Date (`discovery_date`), Flags (`flags`), IDs (`ids`), Involvements (`involvements`),
```

Comment:
I think we should leave this as awe (singular) as it is an enumeration and as such we can always provide more than one matching weakness. This impacts all these `cwe` &rightarrow; `cwes` changes.

Reply:
**Part 1 (Common Weakness Enumerations (CWEs) => Common Weakness Enumeration (CWE)):** Done in 5908da13

**Part 2 (`cwes` => `cwe`):** I disagree. IMHO we should use the principle of least surprise: All other properties use plural, if they contain a list and singular if it is just one entry. Therefore, it would be inconsistent, if we use `cwe` here.

Reply:
**Part 2 (`cwes` => `cwe`):** I disagree. IMHO we should use the principle of least surprise: All other properties use plural, if they contain a list and singular if it is just one entry. Therefore, it would be inconsistent, if we use `cwe` here.

Reply:
Well, for an array that might be true (for those more used to REST terms than folders) but when using an object as proposed in my feedback it is much clearer to access the members per `cwe['42']` even (when extending to a prefix stripped indexing) otherwise still the `cwe['CWE-42']` reads very natural for me.

When mapping eg. to a python dict (from a json.load) you could iterate over that directly as per:

```python
for number, data in cwe.items():
    cwe_id = f'CWE-{number}'
    # ...
```

If we really want to indicate the "one or more such things" property, the term `cwes` is weak (in my reading) as a lower-cased acronym with a plural "s" always reads weird (and wrong).

Maybe we could spend some more letters?

If it must be an array (I doubt that) it could be `cwe_list`.

If it can be an object (I prefer) it could be `cwe_map`.

My main goal is to map unique identifiers we reuse into identifiers in our structures instead of ignoring that property and injecting extra layers for "one or more" as an array of objects is.

Reply:
I see the intend.

My concern come from the user-side: If you want to display a CSAF in your website (without writing additional code) today, you can use logicless templates (e.g. [mustache.js](https://github.com/janl/mustache.js/)). This allows you to display JSON data as rendered HTML.
If we use an array, the transition is as simple as:

CSAF 2.0
```
CWE: {{cwe.id}} - {{cwe.name}}
```
CSAF 2.1
```
CWEs: {{#cwes}} {{id}} - {{name}} ({{version}}) {{/cwes}}
```

Using the `pattern_properties` would require to implement a new function as the key is no longer fixed (but contains information).